AdnuntiusSDK/AdnuntiusAdWebView.swift

```diff
@@ -179,51 +176,112 @@ public class AdnuntiusAdWebView: WKWebView, WKUIDelegate, WKNavigationDelegate,
      */
-    @objc open func loadAd(_ config: AdRequest, completionHandler: AdLoadCompletionHandler? = nil, adnSdkHandler: AdnSdkHandler? = nil) -> Bool {
-        if self.handlerAdaptor == nil {
-            self.handlerAdaptor = HandlerAdaptor(self)
-        }
-        self.handlerAdaptor!.adnAdLoadHandler = completionHandler
-        self.handlerAdaptor!.adnSdkHandler = adnSdkHandler
-        setupCallbacks(self.handlerAdaptor!)
-        
-        let request = AdUtils.toJson(config, self.env)
-        self.loadHTMLString(request.script, baseURL: request.baseUrl)
-        return true
+    @objc open func loadAd(_ request: AdRequest, _ handler: LoadAdHandler) {
+        self.loadAd(request, handler, delayViewEvents: true)
     }
     
+    /**
+     Parameter config: The Ad Request configuration
+     Parameter loadAdHandler:
+     Parameter: delayViewEvents: if true will delay visible and viewable events until updateView(...) is called.  If false will generate visible and viewable events as soon as the ad is rendered.
+    */
+    @objc open func loadAd(_ request: AdRequest, _ handler: LoadAdHandler, delayViewEvents : Bool) {
+        self.handlerAdaptor.adnAdLoadHandler = nil
+        self.handlerAdaptor.adnSdkHandler = nil
+        setupCallbacks(delayViewEvents, handler)
+
+        let requestJson = AdUtils.toJson(self.adId, request, self.env, self.logger.debug)
+        self.logger.verbose("\(requestJson)")
+        self.loadHTMLString(requestJson.script, baseURL: requestJson.baseUrl)
+    }
+
     /*
-     Swift Only implementation which relies on default implementations of methods in the LoadAdHandler for extensibility
+     This method is used by calling code to let the AdnuntiusWebView know if its currently
+     visible in the device view port.
      */
-    open func loadAd(_ config: AdRequest, _ loadAdHandler: LoadAdHandler) -> Void {
-        setupCallbacks(loadAdHandler)
-        let request = AdUtils.toJson(config, self.env)
-        self.loadHTMLString(request.script, baseURL: request.baseUrl)
+    @objc open func updateView(_ scrollView: UIScrollView) {
+        if (self.adId == nil) {
+            self.logger.error("loadViewAd does not support delaying viewable / visible events - use loadAd")
+            return
+        }
+
+        let container = CGRect(origin: scrollView.contentOffset, size: scrollView.frame.size)
+        // for a table view we need the superview for correct visiblity calculations
+        if scrollView is UITableView {
+            // a superview will not be accessible if the cell is not on screen, is a good thing
+            if let superview = self.superview {
+                self.logger.verbose("container: \(container)")
+                self.logger.verbose("superview frame: \(superview.frame)")
+                let percentage = RectUtils.percentageContains(container, superview.frame)
+                updateView(percentage)
+            }
```

Comment:
Simplification which also allows different order of subviews on the viewStack.

`        if scrollView is UITableView {
            let percentage = RectUtils.percentageContains(container, convert(self.frame, to: scrollView))
            updateView(percentage)
        } `